### PR TITLE
refactor model parsing

### PR DESCRIPTION
**Pull request by mxyng**
- refactor metadata

### CODE REVIEW

File: server/images.go [hunk 1]
```diff
@@ -53,7 +51,7 @@ type Model struct {
 }
 
 func (m *Model) IsEmbedding() bool {
-	return slices.Contains(m.Config.ModelFamilies, "bert") || slices.Contains(m.Config.ModelFamilies, "nomic-bert")
```
Comment (mxyng) [inline server/images.go:56]: TODO

File: server/routes.go [hunk 2]
```diff
@@ -715,11 +715,11 @@ func GetModelInfo(req api.ShowRequest) (*api.ShowResponse, error) {
 
 	modelDetails := api.ModelDetails{
 		ParentModel:       model.ParentModel,
-		Format:            model.Config.ModelFormat,
-		Family:            model.Config.ModelFamily,
-		Families:          model.Config.ModelFamilies,
-		ParameterSize:     model.Config.ModelType,
-		QuantizationLevel: model.Config.FileType,
+		// Format:            model.Config.ModelFormat,
```
Comment (mxyng) [inline server/routes.go:718]: TODO

File: server/routes.go [hunk 3]
```diff
@@ -788,11 +788,11 @@ func ListModelsHandler(c *gin.Context) {
 		}
 
 		modelDetails := api.ModelDetails{
-			Format:            model.Config.ModelFormat,
-			Family:            model.Config.ModelFamily,
-			Families:          model.Config.ModelFamilies,
-			ParameterSize:     model.Config.ModelType,
-			QuantizationLevel: model.Config.FileType,
+			// Format:            model.Config.ModelFormat,
```
Comment (mxyng) [inline server/routes.go:791]: TODO

File: llm/ggml.go [hunk 4]
```diff
@@ -171,7 +279,9 @@ func DecodeGGML(r io.ReadSeeker) (*GGML, error) {
 	}
 
 	model, err := c.Decode(&ro)
-	if err != nil {
+	if errors.Is(err, io.EOF) {
+		// noop
```
Comment (BruceMacD) [inline llm/ggml.go:283]: What is the case that causes this error, and why is it ok?

File: server/images.go [hunk 5]
```diff
@@ -160,12 +185,12 @@ func GetModel(name string) (*Model, error) {
...
 
 	model := &Model{
+		Digest:    digest,
 		Name:      mp.GetFullTagname(),
 		ShortName: mp.GetShortTagname(),
-		Digest:    digest,
 		Template:  "{{ .Prompt }}",
-		License:   []string{},
 		Size:      manifest.GetTotalSize(),
+		kvs:       make(map[string]llm.KV),
```
Comment (BruceMacD) [inline server/images.go:193]: this is a map of strings to maps of strings, which is a bit hard on the brain for me, any ideas to flatten this out a bit? What does `metadata.Ref` represent here?

Comment (mxyng) [reply server/images.go:193]: it's a map of `metadata.Ref` which is the model this layer references to its key value. here's a concrete example

```go
kvs := map[string]llm.KV{
  "sha256:170370233dd5c5415250a2ecd5c71586352850729062ccef1496385647293868": map[string]any{
    "general.architecture": "llama",
    "general.file_type": 2,
    "general.parameter_count": 7241732096,
    "general.quantization_version": 2,
    ...
  }
}
```

File: llm/ggml.go (hunk 4 above)
Comment (mxyng) [reply llm/ggml.go:283]: Unlike GGUF which knows how much it needs to read, GGLA reads until EOF at which point it returns

File: llm/llm.go [hunk 6]
```diff
@@ -53,15 +53,14 @@ func New(model string, adapters, projectors []string, opts api.Options) (LLM, er
 	size := ggml.Size
 
 	// fp16 k,v matrices require = n_ctx * n_layer * n_embd / n_head * n_head_kv * 2 bytes each * 2 key and value
-	kv := 2 * 2 * int64(opts.NumCtx) * int64(ggml.NumLayers()) * int64(ggml.NumEmbed()) * int64(ggml.NumHeadKv()) / int64(max(ggml.NumHead(), 1))
+	kv := 2 * 2 * int64(opts.NumCtx) * int64(ggml.KV().BlockCount()) * int64(ggml.KV().EmbeddingLength()) * int64(ggml.KV().HeadCountKV()) / int64(ggml.KV().HeadCount())
```
Comment (BruceMacD) [inline llm/llm.go:56]: there is a potential divide by 0 error here that didnt used to be here, although it shouldnt be possible

File: server/images.go [hunk 7]
```diff
@@ -376,16 +420,6 @@ func CreateModel(ctx context.Context, name, modelFileDir string, commands []pars
 					return err
 				}
 
-				// if the model is still not in gguf format, error out
-				if fromConfig.ModelFormat != "gguf" {
```
Comment (BruceMacD) [inline server/images.go:380]: What happens if someone imports old ggml now?

**Review by BruceMacD — COMMENTED** [review]
This looks pretty much ready to me, just the open question about if someone imports an older ggml formatted file

File: server/images.go (hunk 7 above)
Comment (mxyng) [reply server/images.go:380]: llm.Decode will raise an ErrUnsupportedFormat

**Review by BruceMacD — APPROVED** [review]
lets get this one in before there are conflicts, looks good to me

**Review by jmorganca — APPROVED** [review]

**Pull request merged**